### PR TITLE
Implement split_inclusive for slice and str

### PR DESCRIPTION
**Pull request by golddranks**
# Overview
* Implement `split_inclusive` for `slice` and `str` and `split_inclusive_mut` for `slice`
* `split_inclusive` is a substring/subslice splitting iterator that includes the matched part in the iterated substrings as a terminator.
* EDIT: The behaviour has now changed, as per @KodrAus 's input, to the same semantics with the `split_terminator` function. I updated the examples below.
* Two examples below:
```Rust
    let data = "\nMäry häd ä little lämb\nLittle lämb\n";
    let split: Vec<&str> = data.split_inclusive('\n').collect();
    assert_eq!(split, ["\n", "Märy häd ä little lämb\n", "Little lämb\n"]);
```

```Rust
    let uppercase_separated = "SheePSharKTurtlECaT";
    let mut first_char = true;
    let split: Vec<&str> = uppercase_separated.split_inclusive(|c: char| {
        let split = !first_char && c.is_uppercase();
        first_char = split;
        split
    }).collect();
    assert_eq!(split, ["SheeP", "SharK", "TurtlE", "CaT"]);
```

# Justification for the API
* I was surprised to find that stdlib currently only has splitting iterators that leave out the matched part. In my experience, wanting to leave a substring terminator as a part of the substring is a pretty common usecase.
* This API is strictly more expressive than the standard `split` API: it's easy to get the behaviour of `split` by mapping a subslicing operation that drops the terminator. On the other hand it's impossible to derive this behaviour from `split` without using hacky and brittle `unsafe` code. The normal way to achieve this functionality would be implementing the iterator yourself.
* Especially when dealing with mutable slices, the only way currently is to use `split_at_mut`. This API provides an ergonomic alternative that plays to the strengths of the iterating capabilities of Rust. (Using `split_at_mut` iteratively used to be a real pain before NLL, fortunately the situation is a bit better now.)

# Discussion items
* <s>Does it make sense to mimic `split_terminator` in that the final empty slice would be left off in case of the string/slice ending with a terminator? It might do, as this use case is naturally geared towards considering the matching part as a terminator instead of a separator.</s>
  * EDIT: The behaviour was changed to mimic `split_terminator`.
* Does it make sense to have `split_inclusive_mut` for `&mut str`?

### CODE REVIEW

**Comment by Centril** [comment]
cc @bluss @kennytm

**Comment by SimonSapin** [comment]
Hi @Centril. If you’d like help specifically from me feel free to mention me. If you believe highfive’s choice of reviewer is not appropriate (maybe that person is away for a couple weeks?) and need to find another reviewer please consider picking from https://github.com/rust-lang/highfive/blob/master/highfive/configs/rust-lang/rust.json.

r? @sfackler

**Comment by SimonSapin** [comment]
(Why reassign a third time so soon after the second one?)

**Comment by Dylan-DPC** [comment]
Ah didn't see that you reassigned.

**Comment by KodrAus** [comment]
> Does it make sense to mimic split_terminator in that the final empty slice would be left off in case of the string/slice ending with a terminator?

I think this makes sense too, and is how I’d expect this to work.

**Comment by JohnCSimon** [comment]
Ping from triage: @golddranks can you please address the merge conflict?

**Comment by JohnCSimon** [comment]
Pinging again from triage: @golddranks can you please address the merge conflict?

**Comment by golddranks** [comment]
@JohnCSimon Ah, sorry! I'll fix the merge conflict now.

@KodrAus Thanks for the input. I agree. I'll change the behaviour to that and address that in docs.

**Comment by golddranks** [comment]
(Sorry for fixing seemingly stupid errors here, the tests/builds take super long to run locally ATM for some reason, so I'm pushing changes "opportunisticly".)

**Comment by hdhoang** [comment]
ping from triage @KodrAus, could you update your review? thanks

File: src/liballoc/tests/str.rs [hunk 1]
```diff
@@ -1247,6 +1247,49 @@ fn test_split_char_iterator_no_trailing() {
...
+}
+
+#[test]
+fn test_split_char_iterator_inclusive_rev() {
+    let data = "\nMäry häd ä little lämb\nLittle lämb\n";
+
+    let split: Vec<&str> = data.split_inclusive('\n').rev().collect();
+    assert_eq!(split, ["Little lämb\n", "Märy häd ä little lämb\n", "\n"]);
+
+    // Note that the predicate is stateful and thus dependent
```
Comment (KodrAus) [inline src/liballoc/tests/str.rs]: This seems reasonable to me.

**Comment by KodrAus** [comment]
Thanks @golddranks! This looks good to me.

Would you like to squash these commits down and I’ll merge in?

**Review by KodrAus — APPROVED** [review]

**Comment by Centril** [comment]
👋 @KodrAus

**Comment by eggyal** [comment]
It would be nice to be able to collect a `SplitInclusive` back into a slice.

**Pull request merged**